File: 프로그래머스/Level3/베스트앨범.py
# https://programmers.co.kr/learn/courses/30/lessons/42579

def solution(genres, plays):
    dic = dict()
    total_play = dict()
    for i in range(len(genres)):
        if genres[i] in dic:
            dic[genres[i]].append((i, plays[i]))
            total_play[genres[i]] += plays[i]
        else:
            dic[genres[i]] = [(i,plays[i])]
            total_play[genres[i]] = plays[i]
    sorted_arr = sorted(total_play.items(), key=lambda x:x[1],reverse=True)
    answer = []
    for genre, p in sorted_arr:
        temp_arr = dic[genre]
        temp_arr.sort(key=lambda x:x[1], reverse=True)
        temp_cnt = 0
        for num, pp in temp_arr:
            answer.append(num)
            temp_cnt += 1
            if temp_cnt == 2:
                break
    return answer

File: 프로그래머스/Level3/베스트앨범.js
// https://programmers.co.kr/learn/courses/30/lessons/42579?language=javascript

const solution = (genres, plays) => {
  const total_plays = {};
  const each_plays = {};
  genres.forEach((g, idx) => {
    if (g in total_plays) {
      total_plays[g] += plays[idx];
      each_plays[g].push([idx, plays[idx]]);
    } else {
      total_plays[g] = plays[idx];
      each_plays[g] = [[idx, plays[idx]]];
    }
  });
  const total_plays_list = [];
  for (const k in total_plays) {
    total_plays_list.push([k, total_plays[k]]);
  }
  total_plays_list.sort((a, b) => b[1] - a[1]);
  for (const k in each_plays) {
    each_plays[k].sort((a, b) => b[1] - a[1]);
  }
  return total_plays_list.reduce(
    (prev, cur) => [
      ...prev,
      ...each_plays[cur[0]].slice(0, 2).map((item) => item[0]),
    ],
    []
  );
};
